File: haskell/src/Sort/BubbleSort.hs
module Sort.BubbleSort where

import Core.Helper

bubbleSort :: Ord a => [a] -> [a]
bubbleSort xs
  | isSorted xs = xs
  | otherwise   = bubbleSort $ iter xs -- iterate through xs for swaps again
  where
    isSorted :: Ord a => [a] -> Bool
    isSorted [] = True
    isSorted [x] = True
    isSorted (x:y:xs)
      | x > y     = False
      | otherwise = isSorted $ y:xs

    iter :: Ord a => [a] -> [a]
    iter []  = []
    iter [x] = [x]
    iter (x:y:ys)
      | x < y = x:(iter $ y:ys)
      | otherwise = y:(iter $ x:ys)

File: go/src/sort/bubblesort.go
package sort

import "helper"

func BubbleSort(arr []interface{}) (sorted []interface{}, err error) {
  sorted := make([]interface[])
  sorted = copy(sorted, arr)

  // validate that we can sort the array
  if len(sorted) == 0 {
    err := nil
    return
  }

  if !helper.IsOfComparable(arr) {
    err := errors.New("slice elements must be of a comparable type")
    return
  }

  for isSorted, err := helper.IsSorted(sorted); !isSorted {
    if err != nil {
      return
    }

    for i := 0; i < len(sorted) - 1; i++ {
      if sorted[i] > sorted[i + 1] {
        if sorted, err = helper.Swap(sorted, i, i + 1); err != nil {
          // had a problem during the swap call
          return
        }
      }
    }
  }

  err = nil

  return
}
